lxd/metrics.go

```diff
@@ -0,0 +1,113 @@
+package main
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/lxd/metrics"
+	"github.com/lxc/lxd/lxd/response"
+)
+
+var metricsCmd = APIEndpoint{
+	Path: "metrics",
+
+	Get: APIEndpointAction{Handler: metricsGet, AccessHandler: allowProjectPermission("containers", "view")},
+}
+
+// swagger:operation GET /1.0/metrics metrics metrics_get
+//
+// Get metrics
+//
+// Gets metrics of instances.
+//
+// ---
+// produces:
+//   - text/plain
+// parameters:
+//   - in: query
+//     name: project
+//     description: Project name
+//     type: string
+//     example: default
+// responses:
+//   "200":
+//     description: Metrics
+//     schema:
+//       type: string
+//       description: Instance metrics
+//   "403":
+//     $ref: "#/responses/Forbidden"
+//   "500":
+//     $ref: "#/responses/InternalServerError"
+func metricsGet(d *Daemon, r *http.Request) response.Response {
+	var (
+		dbInstances []db.Instance
+		err         error
+	)
+
+	// Return cached metrics if requests are less than 15 seconds apart.
+	if d.metricsLastFetchedTime.Add(15 * time.Second).After(time.Now()) {
+		return response.SyncResponsePlain(true, d.metrics.String())
+	}
+
+	d.metricsLastFetchedTime = time.Now()
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		filter := db.InstanceTypeFilter(instancetype.Any)
+
+		projectName := queryParam(r, "project")
+		if projectName != "" {
+			filter.Project = &projectName
+		}
+
+		dbInstances, err = tx.GetInstances(filter)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances, err := instance.LoadAllInternal(d.State(), dbInstances)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	d.metrics = metrics.NewMetricSet(nil)
+
+	wgAction := sync.WaitGroup{}
+	metricsLock := sync.Mutex{}
+
+	for _, inst := range instances {
+		// Ignore stopped instances
+		if !inst.IsRunning() {
+			continue
+		}
+
+		wgAction.Add(1)
+
+		go func(inst instance.Instance) {
```

Comment:
The locking and `metricsGetBuilding` code makes sense.

Your proposed channel solution however has a problem. `instCount` can be greater than the amount of metrics collected, if `inst.Metrics()` returns an error. Incrementing `instCount` (with locking) inside of the goroutine so that `instCount` equals the number of collected metrics causes a different problem. The `for i := 0; i < instCount; i++` loop might wrongly assume `instCount = 0` and we'll therefore never get any metrics.

Reply:
Yes true, in that case I suggest rather than returning you send nil down the channel and then ignore nils on the merging side.